File: Master_of_Engineering_project/Testing_scripts/Go/vpn_client.go
package main

import (
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"time"
)

const (
	iterations  = 100
	hostingName = "render"
	csvFile     = "wyniki_testu_vpn.csv"
	base        = "https://rendertesting-wdkv.onrender.com/" 
)

// Funkcja mierząca czas
func measureRequest(url string) float64 {
    // Wyłączone Keep-Alive
    tr := &http.Transport{
        DisableKeepAlives: true,
    }
    client := &http.Client{
        Transport: tr,
        Timeout:   15 * time.Second,
    }

    start := time.Now()
    
    // Tworzenie obiektu żądania
    req, err := http.NewRequest("GET", url, nil)
    if err != nil {
        return 0
    }
    
    req.Header.Set("Connection", "close")
    req.Header.Set("Cache-Control", "no-cache")

    resp, err := client.Do(req)
    if err != nil {
        fmt.Printf("\nBłąd zapytania do %s: %v", url, err)
        return 0
    }
    defer resp.Body.Close()
	
    _, _ = io.Copy(io.Discard, resp.Body)

    return time.Since(start).Seconds() * 1000
}

func main() {
	// Otwieramy plik w trybie dopisywania
	file, err := os.OpenFile(csvFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		fmt.Println("Błąd otwarcia pliku CSV:", err)
		return
	}
	defer file.Close()

	stat, _ := file.Stat()
	if stat.Size() == 0 {
		file.WriteString("\xEF\xBB\xBF") // BOM dla Excela
		file.WriteString("sep=;\n")
		file.WriteString("Timestamp;Hosting;Test;1KB;10KB;100KB\n")
	}

	now := time.Now()
	timestamp := now.Format("2006-01-02 15:04:05")

	for i := 1; i <= iterations; i++ {
		t1 := measureRequest(base + "response_1KB")
		t10 := measureRequest(base + "response_10KB")
		t100 := measureRequest(base + "response_100KB")

		val1 := strings.Replace(fmt.Sprintf("%.2f", t1), ".", ",", 1)
		val10 := strings.Replace(fmt.Sprintf("%.2f", t10), ".", ",", 1)
		val100 := strings.Replace(fmt.Sprintf("%.2f", t100), ".", ",", 1)

		row := fmt.Sprintf("%s;%s;Test %d;%s;%s;%s\n",
			timestamp, hostingName, i, val1, val10, val100)

		file.WriteString(row)
	}
	//pusty wiersz
	file.WriteString(";;;;;\n")
	
	fmt.Println("\nWyniki zapisane do pliku:", csvFile)
}

File: Master_of_Engineering_project/Testing_scripts/Node.js/vpn_client.js
const axios = require('axios');
const fs = require('fs');
const path = require('path');
const { performance } = require('perf_hooks');
const https = require('https');

// Konfiguracja
const iterations = 100;
const hostingName = "vercel";
const csvFile = path.join(__dirname, "wyniki_testu_vpn.csv");
const base = "https://vercel-testing-nu-indol.vercel.app/";

const urls = {
    "1KB": base + "response_1KB.js",
    "10KB": base + "response_10KB.js",
    "100KB": base + "response_100KB.js"
};

//Funkcja mierząca całkowity czas transmisji (od startu do pobrania danych)
async function measureRequest(url) {
    // cache buster
    const finalUrl = `${url}?cb=${Date.now()}_${Math.random().toString(36).substring(7)}`;
    const startTime = performance.now();
    try 
    {
        const agent = new https.Agent({ 
            keepAlive: false,
            rejectUnauthorized: false
        });

        const response = await axios.get(finalUrl, { 
            timeout: 15000,
            httpsAgent: agent,
            responseType: 'arraybuffer',
            headers: { 
                'Connection': 'close',
                'Cache-Control': 'no-cache, no-store, must-revalidate',
                'Pragma': 'no-cache' 
            }
        });

        const endTime = performance.now();

        return endTime - startTime; 
    } catch (error) {
        if (error.code === 'ECONNRESET' || error.message.includes('socket hang up')) {
            console.error(`[Hong Kong/VPN?] Połączenie zerwane (Socket Hang Up) dla: ${url}`);
        } else {
            console.error(`Błąd ${url}: ${error.message}`);
        }
        return 0; 
    }
}

async function runTests() {
    const allResults = [];

    for (let i = 1; i <= iterations; i++) {
        const row = {};
        row['1KB'] = await measureRequest(urls['1KB']);
        row['10KB'] = await measureRequest(urls['10KB']);
        row['100KB'] = await measureRequest(urls['100KB']);

        allResults.push(row);
    }

    saveToCsv(allResults);
}

// Funkcja zapisująca wyniki do CSV
function saveToCsv(results) {
    const fileExists = fs.existsSync(csvFile);
    let csvContent = "";

    if (!fileExists) {
        csvContent += "\ufeff"; // BOM dla Excela
        csvContent += "sep=;\n";
        csvContent += "Timestamp;Hosting;Test;1KB;10KB;100KB\n";
    }
    const date = new Date();
    const timestamp = date.toISOString().replace(/T/, ' ').replace(/\..+/, '');

    results.forEach((r, index) => {
        const val1 = r['1KB'].toFixed(2).replace('.', ',');
        const val10 = r['10KB'].toFixed(2).replace('.', ',');
        const val100 = r['100KB'].toFixed(2).replace('.', ',');

        csvContent += `${timestamp};${hostingName};Test ${index + 1};${val1};${val10};${val100}\n`;
    });

    // Pusty wiersz
    csvContent += ";;;;;\n";

    try {
        fs.appendFileSync(csvFile, csvContent, 'utf8');
        console.log(`\nWyniki zapisane do pliku: ${csvFile}`);
    } catch (err) {
        console.error("Błąd zapisu do pliku CSV! ", err);
    }
}

runTests();